CefSharp.Wpf/IME/IMEWpfKeyboardHandler.cs

```diff
@@ -0,0 +1,305 @@
+            // parameters given to ::ImmSetCandidateWindow() with its 'dwStyle'
+            // parameter CFS_CANDIDATEPOS.
+            // Therefore, we do not only call ::ImmSetCandidateWindow() but also
+            // set the positions of the temporary system caret if it exists.
+            var candidatePosition = new NativeIME.CANDIDATEFORM {
+                dwIndex = 0,
+                dwStyle = (int)NativeIME.CFS_CANDIDATEPOS,
+                ptCurrentPos = new NativeIME.POINT(x, y),
+                rcArea = new NativeIME.RECT(0, 0, 0, 0)
+            };
+            NativeIME.ImmSetCandidateWindow(hIMC, ref candidatePosition);
+
+            if (_systemCaret)
+            {
+                NativeIME.SetCaretPos(x, y);
+            }
+
+            if (_languageCodeId == NativeIME.LANG_KOREAN)
+            {
+                // Chinese IMEs and Japanese IMEs require the upper-left corner of
+                // the caret to move the position of their candidate windows.
+                // On the other hand, Korean IMEs require the lower-left corner of the
+                // caret to move their candidate windows.
+                y += kCaretMargin;
+            }
+            // Japanese IMEs and Korean IMEs also use the rectangle given to
+            // ::ImmSetCandidateWindow() with its 'dwStyle' parameter CFS_EXCLUDE
+            // to move their candidate windows when a user disables TSF and CUAS.
+            // Therefore, we also set this parameter here.
+            var excludeRectangle = new NativeIME.CANDIDATEFORM
+            {
+                dwIndex = 0,
+                dwStyle = (int)NativeIME.CFS_EXCLUDE,
+                ptCurrentPos = new NativeIME.POINT(x, y),
+                rcArea = new NativeIME.RECT(rc.X, rc.Y, x, y + kCaretMargin)
+            };
+            NativeIME.ImmSetCandidateWindow(hIMC, ref excludeRectangle);
+
+            NativeIME.ImmReleaseContext(hwnd, hIMC);
+        }
+
+        private void DestroyImeWindow(IntPtr hwnd)
+        {
+            if (_systemCaret)
+            {
+                NativeIME.DestroyCaret();
+                _systemCaret = false;
+            }
+        }
+
+        internal void ChangeCompositionRange(Range selectionRange, List<Rect> bounds)
+        {
+            _compositionBounds = bounds;
+            MoveImeWindow(_source.Handle);
+        }
+
+        private void UpdateCaretPosition(int index)
+        {
+            MoveImeWindow(_source.Handle);
+        }
+
+        [SuppressMessage("Microsoft.Performance", "CA1811:AvoidUncalledPrivateCode")]
+        private static Int32 HIWORD64(IntPtr ptr)
+        {
+            return (Int32)((ptr.ToInt64() >> 16) & 0xFFFFFFFF);
+        }
+
+        [SuppressMessage("Microsoft.Performance", "CA1811:AvoidUncalledPrivateCode")]
+        private static Int32 LOWORD64(IntPtr ptr)
+        {
+            return (Int32)(ptr.ToInt64() & 0xFFFFFFFF);
+        }
```

Comment:
LOWORD64 seems to return the low 32 bits (bits 0 through 31), but HIWORD64 seems to return bits 16 through 47. Is that what you intended? I would have thought you'd want the low 32 and high 32 bits, or the low 16 and next highest 16 bits.